### PR TITLE
[Feature] Episode Progress Tracking and Resume

### PR DESCRIPTION
**Pull request by michielcx**
This pull request introduces a new feature that enhances user experience by allowing them to continue watching an episode from where they left off.

Progress of each episode is stored in IndexedDB using Dexie.js. The progress is saved every second, so even when the application closes unexpectedly, the progress is remembered.
Additionally the episode progress is displayed on Card components using halfmoon's progress bar and kept automatically up-to-date with Dexie's liveQuery.

I made this for myself, but I can imagine others like this feature.

### CODE REVIEW

**Comment by ThaUnknown** [comment]
I don't fuck with your description as it looks AI generated as fuck

I fuck with this code tho, a lot, it's clean, but you really dont need to use dexie.js, localstorage should be more than enough, maybe even consider using cache store so it times out old progress for you? it's definitely not a needed dependency

Media can fail to resolve, you shouldn't track progress when that happens: https://github.com/ThaUnknown/miru/blob/master/common/modules/animeresolver.js#L206
the code for this isn't clean so read at your own risk lol

Do you handle moments when there's no file media? that can sometimes happen when a person plays a file called `myvideo.mkv` Miru won't resolve what media it is, and all the metadata for the now playing media will either use old media, or no media at all

the currently media code is so scuffed it's not even funny, so I'm not sure myself what happens, I haven't time to re-write it

File: common/modules/animeprogress.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import Dexie, { liveQuery } from 'dexie';
+
+export const db = new Dexie('miru');
```
Comment (ThaUnknown) [inline common/modules/animeprogress.js:3]: IDB is overkill for this, the amt of data is tiny

File: common/modules/animeprogress.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  });
+}
+
+// Return an individual episode's record { mediaId, episode, currentTime, safeduration, createdAt, updatedAt }
+export function getAnimeProgress (mediaId, episode) {
+  return db.animeEpisodeProgress.where({ mediaId, episode }).first();
+}
+
+// Set an individual episode's progress
+export async function setAnimeProgress ({mediaId, episode, currentTime, safeduration}) {
```
Comment (ThaUnknown) [inline common/modules/animeprogress.js:39]: this data should expire after say 3 months or so, maybe consider using cache storage?

File: common/views/Player/Player.svelte [hunk 3]
```diff
@@ -198,6 +200,24 @@
     }
   }
 
+  async function loadAnimeProgress () {
+    const mediaId = current?.media?.media?.id
```
Comment (ThaUnknown) [inline common/views/Player/Player.svelte:204]: if current.media.failed don't load.

File: common/views/Player/Player.svelte [hunk 4]
```diff
@@ -198,6 +200,24 @@
...
+    const episode = current?.media?.episode
+    if (mediaId !== media?.media?.id || episode !== media?.episode) return
+    const animeProgress = await getAnimeProgress(mediaId, episode)
+    if (!animeProgress) return
+    const currentTime = Math.max(animeProgress.currentTime - 5, 0) // Load 5 seconds before
+    seek(currentTime - video.currentTime)
+  }
+
+  function saveAnimeProgress () {
```
Comment (ThaUnknown) [inline common/views/Player/Player.svelte:213]: if current.media.failed don't save.

File: common/views/ViewAnime/EpisodeList.svelte [hunk 5]
```diff
@@ -3,6 +3,7 @@
   import { click } from '@/modules/click.js'
   import { getEpisodeNumberByAirDate } from '@/modules/providers/tosho.js'
   import { alRequest } from '@/modules/anilist'
+  import { liveAnimeProgress } from "@/modules/animeprogress";
```
Comment (ThaUnknown) [inline common/views/ViewAnime/EpisodeList.svelte:6]: don't use extensionless import

File: common/components/cards/PreviewCard.svelte [hunk 6]
```diff
@@ -80,6 +83,11 @@
         on:load={() => { hide = false }}
         src={`https://www.youtube-nocookie.com/embed/${media.trailer?.id}?autoplay=1&controls=0&mute=1&disablekb=1&loop=1&vq=medium&playlist=${media.trailer?.id}`}
       /> -->
+      {#if progress && $progress > 0}
```
Comment (ThaUnknown) [inline common/components/cards/PreviewCard.svelte:86]: how does this look?

File: common/components/cards/EpisodePreviewCard.svelte [hunk 7]
```diff
@@ -29,6 +32,11 @@
         {media.duration}m
       {/if}
     </div>
+    {#if progress && $progress > 0}
```
Comment (ThaUnknown) [inline common/components/cards/EpisodePreviewCard.svelte:35]: how does this look? screenshots pls

File: common/components/cards/EpisodeCard.svelte [hunk 8]
```diff
@@ -34,6 +37,11 @@
           {media.duration}m
         {/if}
       </div>
+      {#if progress && $progress > 0}
```
Comment (ThaUnknown) [inline common/components/cards/EpisodeCard.svelte:40]: how does this look? screenshots pls

**Review by ThaUnknown — CHANGES_REQUESTED** [review]
not bad

**Comment by ThaUnknown** [comment]
additionally I considered using anilist anime notes, see notes under medialist: https://anilist.github.io/ApiV2-GraphQL-Docs

this would solve the entire "how to save this locally", and also show data across devices which is important

I thought of using [3y3 encoding](https://github.com/ArjixWasTaken/3y3/blob/main/3y3.js), this would allow to append data to notes [if the user already has their own] without overriding it, while also making sure it doesn't impact anilist's UI since it's not visible

File: common/views/ViewAnime/EpisodeList.svelte (hunk 5 above)
Comment (EnergoStalin) [reply common/views/ViewAnime/EpisodeList.svelte:6]: Eslint enforce when?
```json
{
  "rules": {
    "import/extensions": [
      "error",
      "ignorePackages",
      {
        "js": "always",
        "ts": "always",
        "svelte": "always"
      }
    ]
  },
}
```

**Comment by michielcx** [comment]
Alright Dexie.js is gone, it works with local storage now. I guess cache storage could have worked too but that kind of focuses on caching requests and responses so I think local storage is better suited. Though instead of using expiry the local storage is limited to the latest 1000 episodes (~0.12MB). 

I don't know how to play a local file like `myvideo.mkv` so I haven't tested it. But it compares current.media and media's id and episode to make sure they're the same and truthy. I am guessing that solves some of those problems.

We could make a background task that syncs the changes from local storage to anilist, but that's a future feature I guess?

Here are some screenshots of the progress bar, mind you I am no designer, so feel free to make it look better:

Episode Card:

Episode Card (hovered):

Small Card:

Small Card (hovered):

Episode List:


Let me know if there's something else

**Comment by ThaUnknown** [comment]
hover cards and small card looks bad, I probably wouldn't keep that

File: common/components/cards/SmallCard.svelte [hunk 9]
```diff
@@ -37,6 +39,11 @@
       </div>
     {/if}
     <img loading='lazy' src={media.coverImage.extraLarge || ''} alt='cover' class='cover-img w-full rounded' style:--color={media.coverImage.color || '#1890ff'} />
+    {#if progress && $progress > 0}
```
Comment (ThaUnknown) [inline common/components/cards/SmallCard.svelte:42]: progress is always a store, so the first check does nothing

Comment (ThaUnknown) [inline common/components/cards/SmallCard.svelte:42]: I'd remove this from small cards, it's missleading as users might think its anime watch progress, not episode watch progress

File: common/components/cards/PreviewCard.svelte (hunk 6 above)
Comment (ThaUnknown) [inline common/components/cards/PreviewCard.svelte:86]: I'd remove this from small cards, it's missleading as users might think its anime watch progress, not episode watch progress

File: common/components/cards/EpisodeCard.svelte (hunk 8 above)
Comment (ThaUnknown) [inline common/components/cards/EpisodeCard.svelte:40]: progress is a store, which is always truthy so the first condition is useless

File: common/components/cards/EpisodePreviewCard.svelte (hunk 7 above)
Comment (ThaUnknown) [inline common/components/cards/EpisodePreviewCard.svelte:35]: progress is a store, which is always truthy so the first condition is useless

File: common/modules/animeprogress.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+function saveToLocalStorage(data) {
+  localStorage.setItem('animeEpisodeProgress', JSON.stringify(data))
+  animeProgressStore.set(data)
+}
+
+const animeProgressStore = writable(loadFromLocalStorage())
+
+// Return an object with the progress of each episode in percent (0-100), keyed by episode number
+export function liveAnimeProgress (mediaId){
+  return derived(animeProgressStore, (data) => {
```
Comment (ThaUnknown) [inline common/modules/animeprogress.js:22]: I have no clue what derived stores do, never used them, but it's nice seeing them used :)

File: common/views/Player/Player.svelte [hunk 11]
```diff
@@ -198,6 +200,25 @@
...
+    const mediaId = current?.media?.media?.id
+    const episode = current?.media?.episode
+    if (current?.media?.failed || mediaId !== media?.media?.id || episode !== media?.episode) return
+    if (buffering || paused || video.readyState < 4) return
+    setAnimeProgress({ mediaId, episode, currentTime: video.currentTime, safeduration })
+  }
+  setInterval(saveAnimeProgress, 1000)
```
Comment (ThaUnknown) [inline common/views/Player/Player.svelte:220]: this is a lot of code and operations to run on a 1s interval, especially considering we seek back 5 seconds anyways so the accuracy is low, I'd set this to 15 maybe 30 seconds?

File: common/views/Player/Player.svelte [hunk 12]
```diff
@@ -198,6 +200,25 @@
     }
   }
 
+  async function loadAnimeProgress () {
+    const mediaId = current?.media?.media?.id
+    const episode = current?.media?.episode
+    if (current?.media.failed || mediaId !== media?.media?.id || episode !== media?.episode) return
```
Comment (ThaUnknown) [inline common/views/Player/Player.svelte:206]: this condition is fucking awful, you have 11 branches in a single if statement!!!!!!!!!!!

this is NOT human readable!

at the start of theses functionns I'd add a guard clause
```js
if (!current?.media?.media?.id || !current.media.media.episode || current.media.failed || !media?.media?.id || !media.episode) return
```
and then remove all the subsequent nullish chaining, that will make it x100 more readable

**Review by ThaUnknown — CHANGES_REQUESTED** [review]
address these and I think this is ready to be shipped

**Comment by ThaUnknown** [comment]
also I've noticed your commits use a different email than the account that's the author of this PR, I assume because you don't want torrent related things linked to your main account, feel free to address that if it's an issue, the PR can be 1 commit if you override/squash it, I don't mind

**Pull request merged**